include/albatross/src/indexing/covariance.hpp

```diff
@@ -0,0 +1,138 @@
+
+  Eigen::Index i, j, ci, cj;
+  std::size_t si, sj;
+  for (i = 0; i < n; i++) {
+    si = static_cast<std::size_t>(i);
+    ci = static_cast<Eigen::Index>(x_inds[si]);
+    for (j = 0; j <= i; j++) {
+      sj = static_cast<std::size_t>(j);
+      cj = static_cast<Eigen::Index>(x_inds[sj]);
+      (*C)(ci, cj) = caller(xs[si], xs[sj]);
+      (*C)(cj, ci) = (*C)(ci, cj);
+    }
+  }
+}
+
+} // namespace details
+
+template <
+    typename CovFuncCaller, typename Grouper, typename X, typename Y,
+    typename std::enable_if_t<details::is_valid_grouper<Grouper, X>::value &&
+                                  details::is_valid_grouper<Grouper, Y>::value,
+                              int> = 0>
+inline Eigen::MatrixXd compute_covariance_by_group(CovFuncCaller caller,
+                                                   Grouper grouper,
+                                                   const std::vector<X> &xs,
+                                                   const std::vector<Y> &ys) {
+  Eigen::MatrixXd C(xs.size(), ys.size());
+  const auto y_indexers = group_by(ys, grouper).indexers();
+
+  auto fill_block = [&caller, &xs, &ys, &C](const auto &x_inds,
+                                            const auto &y_inds) {
+    details::fill_covariance_subset(caller, subset(xs, x_inds), x_inds,
+                                    subset(ys, y_inds), y_inds, &C);
+  };
+
+  auto fill_row = [&](const auto &x_key, const auto &x_inds) {
+    for (const auto &y_pair : y_indexers) {
+      fill_block(x_inds, y_pair.second);
+    }
+  };
+
+  const auto x_indexers = group_by(xs, grouper).indexers();
+  async_apply(x_indexers, fill_row);
+  return C;
+}
+
+template <typename CovFuncCaller, typename Grouper, typename X,
+          typename std::enable_if_t<
+              details::is_valid_grouper<Grouper, X>::value, int> = 0>
+inline Eigen::MatrixXd compute_covariance_by_group(CovFuncCaller caller,
+                                                   Grouper grouper,
+                                                   const std::vector<X> &xs) {
+  Eigen::MatrixXd C(xs.size(), xs.size());
+  const auto x_indexers = group_by(xs, grouper).indexers();
+
+  auto fill_row = [&](const auto &x_key, const auto &x_inds) {
+    for (const auto &y_pair : x_indexers) {
+      const auto &y_inds = y_pair.second;
+      if (x_key < y_pair.first) {
+        // Lower triangle case, fill then copy
+        details::fill_covariance_subset(caller, subset(xs, x_inds), x_inds,
+                                        subset(xs, y_inds), y_inds, &C);
+        for (const auto &x_ind : x_inds) {
+          for (const auto &y_ind : y_inds) {
+            C(y_ind, x_ind) = C(x_ind, y_ind);
+          }
+        }
+      } else if (x_key == y_pair.first) {
+        // Diagonal case, fill both triangles directly
+        details::fill_covariance_subset(caller, subset(xs, x_inds), x_inds, &C);
+      }
+    }
+  };
+
+  async_apply(x_indexers, fill_row);
```

Comment:
It would be nice to decouple the calculation strategy (async vs. not) from the accelerations possible by grouping before evaluation.  We may still want to use a thread pool to evaluate covariances that don't benefit from grouping, and conversely we may want to group wisely but not evaluate blocks concurrently.